src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1757,35 +1734,21 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
   return k;
 }
 
-
-// ----------------------------------------------------------------------------
-// Lookup
-
-// Basic find on classes in the midst of being loaded
-Symbol* SystemDictionary::find_placeholder(Symbol* class_name,
-                                           ClassLoaderData* loader_data) {
-  assert_locked_or_safepoint(SystemDictionary_lock);
-  unsigned int name_hash = placeholders()->compute_hash(class_name);
-  return placeholders()->find_entry(name_hash, class_name, loader_data);
-}
-
-
-// Used for assertions and verification only
-// Precalculating the hash and index is an optimization because there are many lookups
-// before adding the class.
-InstanceKlass* SystemDictionary::find_class(Symbol* class_name, ClassLoaderData* loader_data) {
-  assert_locked_or_safepoint(SystemDictionary_lock);
-  #ifndef ASSERT
-  guarantee(VerifyBeforeGC      ||
-            VerifyDuringGC      ||
-            VerifyBeforeExit    ||
-            VerifyDuringStartup ||
-            VerifyAfterGC, "too expensive");
-  #endif
-
-  Dictionary* dictionary = loader_data->dictionary();
-  unsigned int name_hash = dictionary->compute_hash(class_name);
-  return dictionary->find_class(name_hash, class_name);
+// If a class loader supports parallel classloading handle parallel define requests.
+// find_or_define_instance_class may return a different InstanceKlass
+InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_name, Handle class_loader,
+                                                               InstanceKlass* k, TRAPS) {
+  InstanceKlass* defined_k = find_or_define_helper(class_name, class_loader, k, THREAD);
+  // Clean up original InstanceKlass if duplicate or error
+  if (!HAS_PENDING_EXCEPTION && defined_k != k) {
+    // If a parallel capable class loader already defined this class, register 'k' for cleanup.
+    assert(defined_k != NULL, "Should have a klass if there's no exception");
+    k->class_loader_data()->add_to_deallocate_list(k);
+  } else if (HAS_PENDING_EXCEPTION) {
+    assert(defined_k != NULL, "Should not have a klass if there's an exception");
```

Comment:
Should the assert be `defined_k == NULL`? I wonder if we have a test case that covers this branch.

Reply:
Yes, it should be.  That's a cut/paste error on my part.  There is a test for this: runtime/Metaspace/DefineClass.java

Reply:
Changing it to defined_k == NULL found a bug in a previous refactoring of find_or_define_instance_class.  It should have returned NULL for a pending exception rather than the original k klass.  The old code recreated the exception and had:

  // Can't throw exception while holding lock due to rank ordering
  if (linkage_exception() != NULL) {
    THROW_OOP_(linkage_exception(), NULL); // throws exception and returns
  }

but the new code doesn't clear and recreate the exception.